### ISSUE TITLE
Using Athena with Firehose, Glue etc.

### ISSUE THREAD

**Issue by houmark** [aws-amplify/amplify-cli#2949]
I've been trying to get an analytics/stats dashboard developed for our platform which is built on Amplify.

I've tried several attack angles and read countless articles which all try to get to the same end result; Being able to query data in Athena using ANSI SQL to create stats/charts.

Our data is stored in DynamoDB and I'm hoping for some pointers on best practice, and maybe some boilerplate code to reliably get data from DynamoDB to S3 on create/update. I've played with some Python code, but I cannot really figure out how to get that into S3 in Parquet format, and saving it to JSON in S3 and then have a Glue job pick that up and convert to Parquet seems unnecessarily complex.

The goal is a near-real-time mirror of some of our DynamoDB data in Parquet format in S3 that can then be queried through the Athena API. I read one article that described how to create custom GraphQL queries to query Athena and that seems very promising, but my main struggle right now really is to get data out of DynamoDB and into S3 in Parquet format for optimized storage and query performance.

I saw that Kinesis Firehose features were added to Amplify recently, but I am unsure if those additions will be helpful to this challenge in any way?

If anyone has succeeded in this and would share their adventure of if AWS devs have information in relation to this challenge, I'd appreciate it very much.

**What AWS Services are you utilizing?**

Amplify.

**Comment by davidbiller** [comment]
A few useful infos to understand what is being tried
https://aws.amazon.com/de/blogs/mobile/visualizing-big-data-with-aws-appsync-amazon-athena-and-aws-amplify/
So the part to get the data on a s3 bucket at all is missing here

**Comment by davidbiller** [comment]
So @houmark 
If you don't want to use it as realtime, the AWS DataPipeline makes it quite good -  exporting DynamoDB to S3.

we could talk about this biller@aivy.app

**Comment by RossWilliams** [comment]
Some more information about your requirements would help find a solution. Can you define an acceptable lag in data, number of records/metrics/documents per day, how much historical information you need, whether it is time series data, do you need to do extensive processing of the data or combine it with other data sources/tables, support of ad-hoc queries? There are so many AWS choices, and from your description i'm not sure S3/parquet and Athena is the right direction.

As for Amplify, you could deploy some or all of these pieces inside the same nested stack, but it would be either heavily modifying a function's cloudformation template, or defining them as custom resource in the api/xxx/stacks directory.

I've added analytics, reporting, used firehose, transformed data for time series from amplify data sources, but near 100% of the work is outside the scope of Amplify.

**Comment by houmark** [comment]
Hi @RossWilliams — thanks for your feedback.

I've been on and off researching and trying out different options on this for more than two weeks. Nothing really worked, even just getting a Glue / ETL job to run was creating problems.

So that's when I decided to post here to see if anyone else had something to contribute, without going into too many details.

Less than an hour after posting this, I came across some new Athena features called [Athena Query Federation](https://github.com/awslabs/aws-athena-query-federation) which looks very promising. If you have not read about it yet, this enables you, in a few simple steps, to connect Athena with DynamoDB and 10 other sources, or even code your own connector.

I test-drove it and the overall impression is good, though I did hit a few bugs on the DynamoDB connector, which is expected since this is still only in **Public Preview**. I've reported that and hoping that those issues are solved soon.

As the Athena queries are partly working, I am now trying to wire in GraphQL/AppSync to be able to query Athena using the concept from [here](https://github.com/aws-samples/aws-appsync-visualization-with-athena-app).

Any feedback is greatly appreciated.

**Comment by RossWilliams** [comment]
@houmark The Athena Query Federation uses scans quite heavily, which is why I asked about your dataset size and usage pattern. The cost could easily explode. I thought about recommending other AWS services, but they are very use-case dependent.

**Comment by houmark** [comment]
Yes, I'm aware of that, but the dataset is fairly small at this time (below 10k records in the main tables but that will increase over time of course).

The plan is to try to save the query ID and then attempt to load the spilled result in a Lambda function directly from S3 if the result is based on stale/history data, to avoid repeating queries that will give the same result. Not sure I'll get there in v1, but both price and 5-8 second scan time is a concern which I could solve by doing that.

**Comment by houmark** [comment]
Thanks for pointing me towards Redshift again. I've discarded it for now mainly because I got the understanding that it was even more expensive. I'll give it a second look!

Can I API to Redshift in a similar way that it seems possible with Athena for AppSync?

**Comment by davidbiller** [comment]
@houmark 
As I understood it, you simply want to build a dashboard across tables. Probably for KPI measurements. 

I'm also interested in exactly what and how to use it.

**Comment by houmark** [comment]
@davidbiller Yeah, my main goal is a fairly simple stats dashboard with KPI's on the most critical business processes etc.

Besides a few completely blocking bugs in the Athena Query Federation DynamoDB connector, I have already been able to run some queries that return data that would be very easy to consume from my React web app and draw different charts. I'm hoping the Query Federation team will solve those bugs and while I wait for that, I am doing some data migration in my DynamoDB database to have all the data needed and better prepped in some cases. Once that's done I will try to build the client-side custom query/mutation/subscription setup for AppSync and it's there I want to try to make it use the pipeline Lambda function to shortcut to the results in the spill bucket, which should speed things up and avoid repeating queries and keep costs down.

The entire part about relying on a data lake and data to be copied and ready and having a (hopefully) complete copy of the data is a turn off for me and the way the Query Federation works feels very attractive because it uses the data from the DynamoDB in real-time. Configuring it was easy and fast, so for smaller solutions I feel like this is the best option. Time will tell if that's really the case.

**Comment by davidbiller** [comment]
@houmark 
Yes, I'm coming to the same subject now. 
KPI. 

- The option to do this directly via AppSync is also available. However, the performance is certainly not so good, since many scans would have to be done. 

- The option to do this via a DataPipline, then Glue, then Athena is also not so good, because I use DynamoDB via On-Demand. DataPipeline only works with reserved capacity.

- The Athena Query Federation option also seems to run scans. Is the performance okay? 

- The option over a Lambda still remains. I would then create another index for each table so that I have more options to sort and filter. Then I would use Lambda to build the logic that I would actually query using SQL.

For me personally a day delay would be ok. Then I could always access a S3 backup. But as I said before, it doesn't give that for on-demand. 

I'm going to have to watch Athena Query Federation.

I'm currently having the Pinpoint data stored on an S3, as it is only stored for 90 days in Pinpoint itself. Why actually? I have a dashboard with my analytics, then after 90 days, I have to rebuild the dashboard myself :D

**Comment by davidbiller** [comment]
@houmark 
Another solution that I also find interesting is to go through customResolver and simply save parts of the KPI beforehand in a separate table. 

If a new user is created, there is a lambda which is called as event and sets the number of users: + 1.

There are customResolvers which do not have to load the item first to increase the count. 

Daily active User you can also run a lambda once a day and save the result in the KPI table before. 

I think this will be the way I test first. An advantage I see in it:
Subscriptions to the table via GraphQL.
And Performance. Less cost.

**Comment by RossWilliams** [comment]
Just to put out the way i've tackled the issue:

Where I know the access patterns and metrics needed (and if you have KPIs you likely have this), I use dynamodb streams to write aggregation data to a new dynamo table, and i write it three times for different resolutions. So I have hourly, daily, and weekly buckets. I use the ADD update expression and keep track of request id GUIDS to make sure I don't double-add. Reads are then pretty simple, each metric is a hash key, and all metrics are sorted, so the frontend can request a week's worth of metrics, which will hit the MetricA#daily hash key with the sort key between END and START timestamps.

Analytics data was harder to define the use-cases, and Pinpoint data never reached dynamodb, so that data is sent into Kinesis Firehose, enriched and transformed with a lambda, and inserted into ES. I could set the Firehose backups to go into an S3 bucket, and have a Lambda trigger for the bucket to then insert metrics into dynamodb. This would let all my metrics show up in one place, but its lower on my priority order. For now reporting is both through Kibana and a custom dashboard.

If I were to do this again, I would look to use CloudWatch metrics, and insert metrics from a dynamodb stream if that were the source, or read from an S3 bucket for any other sources. The main reason this wasn't done is that I generate fairly detailed reports and I couldn't work out how to do it without having a cost explosion with the way dimensions affect metric pricing.

**Comment by RossWilliams** [comment]
@davidbiller no need to create custom lambdas if your metrics are on mutations, as these can come through dynamodb streams. Lambda resolvers have an additional cost and speed issue.

**Comment by davidbiller** [comment]
> @davidbiller no need to create custom lambdas if your metrics are on mutations, as these can come through dynamodb streams. Lambda resolvers have an additional cost and speed issue.

mhhh okay but for the ADD update expression i need a custom resolver?!

**Comment by RossWilliams** [comment]
for my setup, all metrics operations (ADD update expressions) are done in DynamoDB streams Lambda triggers. I'm not updating metrics in the same operation that does the actual mutation. This gives me faster mutation responses, separates concerns, and the lambda triggers receives batches of updates for cost efficiency. Downside is my metrics are not realtime, but i can configure the amount of lag.

**Comment by davidbiller** [comment]
> for my setup, all metrics operations (ADD update expressions) are done in DynamoDB streams Lambda triggers. I'm not updating metrics in the same operation that does the actual mutation. This gives me faster mutation responses, separates concerns, and the lambda triggers receives batches of updates for cost efficiency. Downside is my metrics are not realtime, but i can configure the amount of lag.

Ah, now I understand you. Yes, as a lambda trigger is a good way. 
So in the lambda, you use graphql to increase/decrease numbers? 
Else I guess I will lose data, if I access to dynamoDB, read the data, and update. 
So this Is why I said that I want use the custom resolver. 

Mutation -> lambda trigger -> graphQL custom query to increase the number without read it before.

**Comment by houmark** [comment]
Prepping stats on runtime for later use really is an evil path to get on imho.

In my case, one of the basic features is the possibility to select any two dates and have results based on that. Prepping stats for that on beforehand is near impossible.

If a new KPI is developed or the calculations for an existing one is tweaked, then some migration code is needed which is time-consuming. Working to cache stale data results seems a much better investment of my time to be honest.

I understand that some people out there have so much data that it's impossible to query on demand, but then at least those could be prepped and queried behind the scenes or run based on events or time interval, so that the results are already ready when consumed.

I still appreciate this discussion and the different angles you guys have taken or considered.

Also, the Athena DynamoDB Connector does support parallel scans, which should speed up most results as mentioned in the docs:
> The Athena DynamoDB Connector does support parallel scans and will attempt to push down predicates as part of its DynamoDB queries. A hash key predicate with X distinct values will result in X Query calls to DynamoDB. All other predicate scenarios will results in Y number of Scan calls where Y is heuristically determined based on the size of your table and its provisioned throughput.

In my very early testing and without a large dataset, results have been consistent even when my queries have become more advanced. I do seem to be scanning very little data, in most cases, only hundreds of kilobytes, and here Athena is charging a minimum of 10MB, so to optimize the cost, a v2 feature I am considering is combining multiple queries to query a dataset close to or above 10MB and then split up that dataset to be able to consume parts of the results later on (using the v1.5 cache engine I'm considering).

I read a very interesting article taking that angle (on a data lake stored in S3), and they apparently got cost for stats querying down to $0.02 per user/monthly which seems pretty good to me.

**Comment by RossWilliams** [comment]
> So in the lambda, you use graphql to increase/decrease numbers?

No, I call dynamodb directly.
Mutation -> vtl updates dynamo -> dynamo stream trigger -> lambda trigger handler. -> dynamo put/(update expression).

> In my case, one of the basic features is the possibility to select any two dates and have results based on that. Prepping stats for that on beforehand is near impossible.

This is exactly what I attempted with my approach. My sort key is the metric bucket timestamp and I can query using the Between key condition. If I want to reduce it to a single metric, it would be a sum of the results. Only condition is the time resolution is limited by the metric bucket.

I can understand the frustration of creating a new metric and backfilling the results, it takes far more time than writing a new SQL statement. But once you setup the code to do it the first time, there is a lot of reuse. I certainly wouldn't suggest my approach for ad-hoc queries or as a general BI tool, its more suited to a top 10 KPIs situation with hourly monitoring.

I'm very intrigued by the Athena connector, hopefully it will get out of preview soon, and it looks to be limited to US-EAST-1 right now?

**Comment by houmark** [comment]
Yeah, I considered the between query with a key, but I know I will also have to filter more advanced in some situations. Before I even started considering Athena + data lakes etc., I was trying out queries on DynamoDB directly, but the lack of aggregation simply kills any viable route.

I'm not sure if this is only us-east-1, but we are luckily in that region, so I was able to get somewhat started. Please note that I did encounter two major blocking bugs:

1) Which affects all Amplify/AppSync created DynamoDB tables, because tables are created with hyphens and the connector does not map that correctly, resulting in a NotFound error when using a Glue crawler. A glue crawler is not strictly needed, as the connector can do what they call "schema inference" and map fields automatically. Though, in my testing, some fields (especially fields that does not have data on all records in DynamoDB) was not mapped at all, making it impossible to query on them. I reported that and they seem to have a few simple attack angles to solve that.

2) Struct fields seems to choke in some situations for one of my tables. Also reported, not sure how easy the fix is. I do not have plans to query on those struct fields, so if 1 is solved, I may be able to remove those fields in the Glue crawler mapping and be able to query the table without problems.

I've really struggled with this project, so I'm hoping the above is solved fast, so I can move on and get a v1 dashboard up and running.

The long term opportunities here seem promising, as it seems possible for the Amplify CLI to provision this Athena using the connector and by that wire in all the needed parts directly from the CLI.

**Comment by rrrix** [comment]
Hey @houmark, check out this library: https://github.com/awslabs/aws-data-wrangler 

It's a Python framework that you can use to *_very_* easily put Parquet formatted data into S3 from... well anywhere :) 

From reading the thread, I think you'll have to take a step back and rework a lot of your design, but going back to your original post, you wanted to push Parquet into S3 so you can efficiently query with Athena. Using aws-data-wrangler, it wouldn't be much of a stretch to do something like:

1. DynamoDB Streams -> Invoke a Lambda Function to stream/put real time data to S3 in JSON
1. Daily (hourly?) scheduled Lambda Function to "parquetify" your JSON/CSV for efficient Athena queries. 

Athena can query both Parquet (historical data) and JSON (recent data) from a single database/schema, which is why this works. 

As @davidbiller said: 
> So the part to get the data on a s3 bucket at all is missing here

So that's what this will help you do!

**Comment by davidbiller** [comment]
> Hey @houmark, check out this library: https://github.com/awslabs/aws-data-wrangler
> 
> It's a Python framework that you can use to __very__ easily put Parquet formatted data into S3 from... well anywhere :)
> 
> From reading the thread, I think you'll have to take a step back and rework a lot of your design, but going back to your original post, you wanted to push Parquet into S3 so you can efficiently query with Athena. Using aws-data-wrangler, it wouldn't be much of a stretch to do something like:
> 
> 1. DynamoDB Streams -> Invoke a Lambda Function to stream/put real time data to S3 in JSON
> 2. Daily (hourly?) scheduled Lambda Function to "parquetify" your JSON/CSV for efficient Athena queries.
> 
> Athena can query both Parquet (historical data) and JSON (recent data) from a single database/schema, which is why this works.
> 
> As @davidbiller said:
> 
> > So the part to get the data on a s3 bucket at all is missing here
> 
> So that's what this will help you do!

@houmark 
Did you test this solution?
Or any other update?

**Comment by r0zar** [comment]
Can we close this since it's not an active issue or feature in progress?
